<a href="https://colab.research.google.com/github/HansAzharr/ML-Material-Screening/blob/main/02_Feature_engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import os
import sys
import ast

## Load file

In [2]:
base_project_path = "/content/drive/MyDrive/Colab Notebooks/Masters Research Project"
data_folder_path = os.path.join(base_project_path, "Data")
model_folder_path = os.path.join(base_project_path, "Models")

In [3]:
file_path = '/content/drive/MyDrive/Colab Notebooks/Masters Research Project/Data/battery_materials_processed_v2.csv'
if os.path.exists(file_path):
    df = pd.read_csv(file_path)
    print("File loaded successfully.")
else:
    print("File not found at the specified path.")

File loaded successfully.


In [4]:
df.head()

material_id  energy_above_hull           elements   density formula_pretty  \
0  mp-1002568           0.071550  ['Mg', 'Mn', 'O']  4.112037        MgMn2O4   
1    mp-10033           0.000000        ['Nb', 'S']  4.203264           NbS2   
2  mp-1003637           0.029379  ['Li', 'Mn', 'O']  4.350043        LiMn3O6   
3  mp-1003854           0.151309   ['K', 'Mn', 'O']  3.781312          KMnO2   
4  mp-1004375           0.042391  ['Mg', 'Mn', 'O']  4.096519        MgMn4O8   

   nelements      volume  band_gap  is_metal  total_magnetization  ...  \
0        3.0  160.058626    0.6493     False             5.997778  ...   
1        2.0  124.077397    0.0000      True             1.727667  ...   
2        3.0  102.208625    0.3371     False            10.000594  ...   
3        3.0  221.390103    1.3322     False            15.999967  ...   
4        3.0  150.812818    0.2766     False            14.000303  ...   

   theoretical_capacity_vol energy_grav   energy_vol  num_steps  \
0                515.367052  -33.388634  -141.997174        1.0   
1                698.540768  396.561540  1694.835288        2.0   
2                213.454309  115.696156   499.827361        1.0   
3                331.113426    1.188689     3.880221        1.0   
4                829.097077 -219.544732  -787.840537        1.0   

   max_voltage_step  stability_charge  stability_discharge   e_fermi  nsites  \
0          0.000000          0.071550             0.054910  3.079768      14   
1          0.487941          0.000000             0.000000  3.525981       6   
2          0.000000          0.029379             0.059105  2.608843      10   
3          0.000000          0.151309             0.158791  1.706243      16   
4          0.000000          0.042391             0.171669  2.756918      13   

   crystal_system  
0      Monoclinic  
1       Hexagonal  
2      Monoclinic  
3      Tetragonal  
4    Orthorhombic  

[5 rows x 25 columns]

In [5]:
# convert elements string into python list

if 'elements' in df.columns:
    # Check if conversion is needed (i.e., if the first non-null element is a string)
    if not df['elements'].empty and isinstance(df['elements'].iloc[0], str):
        df['elements'] = df['elements'].apply(ast.literal_eval)
        print("Converted 'elements' column from string to list.")
    else:
        print("'elements' column is already in list format or is empty.")
else:
    print("Warning: 'elements' column not found in the DataFrame.")

Converted 'elements' column from string to list.


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3709 entries, 0 to 3708
Data columns (total 25 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   material_id                3709 non-null   object 
 1   energy_above_hull          3709 non-null   float64
 2   elements                   3709 non-null   object 
 3   density                    3709 non-null   float64
 4   formula_pretty             3709 non-null   object 
 5   nelements                  3709 non-null   float64
 6   volume                     3709 non-null   float64
 7   band_gap                   3709 non-null   float64
 8   is_metal                   3709 non-null   bool   
 9   total_magnetization        3709 non-null   float64
 10  theoretical                3709 non-null   bool   
 11  battery_type               3709 non-null   object 
 12  working_ion                3709 non-null   object 
 13  average_voltage            3709 non-null   float

In [7]:
df.head()

material_id  energy_above_hull     elements   density formula_pretty  \
0  mp-1002568           0.071550  [Mg, Mn, O]  4.112037        MgMn2O4   
1    mp-10033           0.000000      [Nb, S]  4.203264           NbS2   
2  mp-1003637           0.029379  [Li, Mn, O]  4.350043        LiMn3O6   
3  mp-1003854           0.151309   [K, Mn, O]  3.781312          KMnO2   
4  mp-1004375           0.042391  [Mg, Mn, O]  4.096519        MgMn4O8   

   nelements      volume  band_gap  is_metal  total_magnetization  ...  \
0        3.0  160.058626    0.6493     False             5.997778  ...   
1        2.0  124.077397    0.0000      True             1.727667  ...   
2        3.0  102.208625    0.3371     False            10.000594  ...   
3        3.0  221.390103    1.3322     False            15.999967  ...   
4        3.0  150.812818    0.2766     False            14.000303  ...   

   theoretical_capacity_vol energy_grav   energy_vol  num_steps  \
0                515.367052  -33.388634  -141.997174        1.0   
1                698.540768  396.561540  1694.835288        2.0   
2                213.454309  115.696156   499.827361        1.0   
3                331.113426    1.188689     3.880221        1.0   
4                829.097077 -219.544732  -787.840537        1.0   

   max_voltage_step  stability_charge  stability_discharge   e_fermi  nsites  \
0          0.000000          0.071550             0.054910  3.079768      14   
1          0.487941          0.000000             0.000000  3.525981       6   
2          0.000000          0.029379             0.059105  2.608843      10   
3          0.000000          0.151309             0.158791  1.706243      16   
4          0.000000          0.042391             0.171669  2.756918      13   

   crystal_system  
0      Monoclinic  
1       Hexagonal  
2      Monoclinic  
3      Tetragonal  
4    Orthorhombic  

[5 rows x 25 columns]

## Filter working ions

In [8]:
df = df.copy()
print(f"Starting Feature Engineering with {len(df)} rows.")

Starting Feature Engineering with 3709 rows.


In [9]:
initial_rows_before_ion_filter = len(df)
allowed_ions = ['Li', 'Na', 'Mg']
df = df[df['working_ion'].isin(allowed_ions)].copy()
print(f"Filtered to keep only working_ion in {allowed_ions}. New row count: {len(df)} (Dropped {initial_rows_before_ion_filter - len(df)} rows).")


Filtered to keep only working_ion in ['Li', 'Na', 'Mg']. New row count: 3445 (Dropped 264 rows).


## Derived Host Material Features

In [10]:
# Define the path to PubChem CSV
# Data is retrieved from National Center for Biotechnology Information (2025).
# https://pubchem.ncbi.nlm.nih.gov/periodic-table/

pubchem_csv_filepath = '/content/drive/MyDrive/Colab Notebooks/Masters Research Project/Data/PubChemElements_all.csv'

property_column_map = {
    'atomic_number': 'AtomicNumber',
    'atomic_mass': 'AtomicMass',
    'electronegativity': 'Electronegativity',
    'covalent_radius': 'AtomicRadius' # Using AtomicRadius as a proxy here
}


In [11]:
element_properties = {}

try:
    # Load the CSV file, select relevant columns, and set 'Symbol' as index
    required_cols = ['Symbol'] + list(property_column_map.values())
    df_elements = pd.read_csv(pubchem_csv_filepath, usecols=required_cols, index_col='Symbol')

    # Convert relevant columns to numeric, coercing errors to NaN
    for csv_col in property_column_map.values():
         df_elements[csv_col] = pd.to_numeric(df_elements[csv_col], errors='coerce')

    # Convert the DataFrame to a dictionary of dictionaries and map column names
    element_properties = df_elements.rename(columns={v: k for k, v in property_column_map.items()}).to_dict('index')

    print(f"Successfully loaded properties for {len(element_properties)} elements from '{pubchem_csv_filepath}'.")
    print("Note: 'AtomicRadius' from the CSV is being used as the 'covalent_radius' feature.")

except FileNotFoundError:
    print(f"Error: The PubChem CSV file was not found at '{pubchem_csv_filepath}'. Please check the path and ensure it's mounted.")
    element_properties = {}
    print("Element properties dictionary is empty due to file not found. Subsequent feature engineering will be affected.")
except Exception as e:
    print(f"An error occurred while processing the PubChem CSV: {e}")
    element_properties = {}
    print("Element properties dictionary is empty due to an error during CSV processing. Subsequent feature engineering will be affected.")

Successfully loaded properties for 118 elements from '/content/drive/MyDrive/Colab Notebooks/Masters Research Project/Data/PubChemElements_all.csv'.
Note: 'AtomicRadius' from the CSV is being used as the 'covalent_radius' feature.


In [12]:
# Derived Host Material Features (14 features)

if 'elements' in df.columns:
    # Safely convert string representation of lists to actual lists
    # Use literal_eval for safer evaluation if the format is consistent with Python literals
    df['elements_parsed'] = df['elements'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else x)

    # Function to get properties and calculate aggregations for a list of elements
    def calculate_elemental_aggregations(element_list):
        if not isinstance(element_list, list):
            return pd.Series([np.nan] * (4 * 3 + 1)) # Return Series with NaN for all 13+1 features if input is not a list

        results = {}
        props_dict = {}
        for prop_name in ['atomic_number', 'atomic_mass', 'electronegativity', 'covalent_radius']:
             # Use .get() with a default of {} to avoid KeyError if element_properties is incomplete
            props = [element_properties.get(el, {}).get(prop_name) for el in element_list if el in element_properties]
            props_dict[prop_name] = pd.Series(props) if props else pd.Series([np.nan]) # Store as Series for easy aggregation

        # Calculate aggregations
        results['num_unique_elements'] = len(set(element_list)) if element_list else 0 # Count unique elements

        for prop_name in ['atomic_number', 'atomic_mass', 'electronegativity', 'covalent_radius']:
            results[f'avg_{prop_name}'] = props_dict[prop_name].mean()
            results[f'min_{prop_name}'] = props_dict[prop_name].min()
            results[f'max_{prop_name}'] = props_dict[prop_name].max()

        # Add the 14th feature: std_electronegativity
        results['std_electronegativity'] = props_dict['electronegativity'].std() if len(props_dict['electronegativity'].dropna()) > 1 else 0.0 # Handle single element case


        return pd.Series(results)

    # Apply the function to create the new features
    aggregated_features = df['elements_parsed'].apply(calculate_elemental_aggregations)

    # Concatenate the new features to the original DataFrame
    df = pd.concat([df, aggregated_features], axis=1)

    print("Created 14 Host Material Features (num_unique_elements, avg/min/max for atomic number, atomic mass, electronegativity, covalent radius, and std_electronegativity).")

    df.drop(columns=['elements_parsed'], inplace=True) # Drop temporary column

else:
    print("Warning: 'elements' column not found. Skipping Host Material Feature engineering.")

Created 14 Host Material Features (num_unique_elements, avg/min/max for atomic number, atomic mass, electronegativity, covalent radius, and std_electronegativity).


## Intercalating Ion Features (Li, Na, Mg specific)

In [13]:
# Predefined properties for common working ions
# Values are sourced from standard periodic table data.

ion_properties = {
    'Li': {'ionic_radius': 0.76, 'atomic_weight': 6.941, 'charge': 1},
    'Na': {'ionic_radius': 1.02, 'atomic_weight': 22.990, 'charge': 1},
    'Mg': {'ionic_radius': 0.72, 'atomic_weight': 24.305, 'charge': 2}
}

In [14]:
# Intercalating Ion Features (5 features)

if element_properties:
    for ion_sym in allowed_ions:
        if ion_sym in element_properties:
            ion_properties[ion_sym]['atomic_number'] = element_properties[ion_sym].get('atomic_number')
            ion_properties[ion_sym]['electronegativity'] = element_properties[ion_sym].get('electronegativity')
        else:
            print(f"Warning: Properties for ion '{ion_sym}' not found in element_properties. Some ion features might be NaN.")
            # Set to NaN if not found to avoid KeyError
            ion_properties[ion_sym]['atomic_number'] = np.nan
            ion_properties[ion_sym]['electronegativity'] = np.nan

if 'working_ion' in df.columns:
    df['ionic_radius_ion'] = df['working_ion'].map(lambda x: ion_properties.get(x, {}).get('ionic_radius'))
    df['atomic_weight_ion'] = df['working_ion'].map(lambda x: ion_properties.get(x, {}).get('atomic_weight'))
    df['charge_ion'] = df['working_ion'].map(lambda x: ion_properties.get(x, {}).get('charge'))
    df['atomic_number_ion'] = df['working_ion'].map(lambda x: ion_properties.get(x, {}).get('atomic_number'))
    df['electronegativity_ion'] = df['working_ion'].map(lambda x: ion_properties.get(x, {}).get('electronegativity'))

    print("Created 5 Intercalating Ion Features (ionic_radius_ion, atomic_weight_ion, charge_ion, atomic_number_ion, electronegativity_ion).")
else:
    print("Warning: 'working_ion' column not found. Skipping Intercalating Ion Features.")


Created 5 Intercalating Ion Features (ionic_radius_ion, atomic_weight_ion, charge_ion, atomic_number_ion, electronegativity_ion).


In [15]:
# One-Hot Encoded categorical Features

if 'working_ion' in df.columns:
    # Define all categorical columns to be one-hot encoded
    columns_to_onehot_encode = ['working_ion']
    if 'crystal_system' in df.columns: # Add crystal_system
        columns_to_onehot_encode.append('crystal_system')

    if 'is_metal' in df.columns and df['is_metal'].dtype == 'bool':
        columns_to_onehot_encode.append('is_metal')
    if 'theoretical' in df.columns and df['theoretical'].dtype == 'bool':
        columns_to_onehot_encode.append('theoretical')

    print("\nApplying One-Hot Encoding to categorical features...")
    df = pd.get_dummies(df, columns=columns_to_onehot_encode, prefix=columns_to_onehot_encode, dummy_na=False, dtype=int)

    # Ensure specific 'working_ion' columns exist and create them if not
    # Prefix will be 'working_ion' as per get_dummies
    for ion in allowed_ions:
        col_name = f'working_ion_{ion}'
        if col_name not in df.columns:
            df[col_name] = 0 # Add column and fill with 0 if no samples of that ion were left after filter

    print(f"Created One-Hot Encoded Features for: {', '.join(columns_to_onehot_encode)}.")
    print(f"Shape after One-Hot Encoding: {df.shape}")
    print("Example of new encoded columns (first 5 rows):")
    # Adjust this print to show columns relevant to your encoded features
    print(df[[col for col in df.columns if any(p in col for p in ['working_ion_', 'crystal_system_', 'is_metal_', 'theoretical_'])]].head())

else:
    print("Warning: 'working_ion' column not found. Skipping One-Hot Encoded Ion Features.")
    # If crystal_system still needs encoding, handle it separately here
    if 'crystal_system' in df.columns:
        print("Warning: 'working_ion' not found, but encoding 'crystal_system' separately.")
        df = pd.get_dummies(df, columns=['crystal_system'], prefix='crystal_system', dummy_na=False, dtype=int)
        print(f"Shape after One-Hot Encoding for crystal_system: {df.shape}")


Applying One-Hot Encoding to categorical features...
Created One-Hot Encoded Features for: working_ion, crystal_system, is_metal, theoretical.
Shape after One-Hot Encoding: (3445, 54)
Example of new encoded columns (first 5 rows):
   theoretical_capacity_grav  theoretical_capacity_vol  working_ion_Li  \
0                 121.181299                515.367052               0   
1                 163.446209                698.540768               1   
2                  49.408746                213.454309               1   
3                 101.435161                331.113426               0   
4                 231.041546                829.097077               0   

   working_ion_Mg  working_ion_Na  crystal_system_Cubic  \
0               0               1                     0   
1               0               0                     0   
2               0               0                     0   
3               1               0                     0   
4               0           

## Drop Redundant Columns

In [16]:
# Drop original columns that have been transformed or replaced
columns_to_drop = ['elements']


if 'crystal_system' in df.columns:
    columns_to_drop.append('crystal_system')

if 'is_metal' in df.columns and df['is_metal'].dtype == 'bool':
    columns_to_drop.append('is_metal')
if 'theoretical' in df.columns and df['theoretical'].dtype == 'bool':
    columns_to_drop.append('theoretical')

df.drop(columns=[col for col in columns_to_drop if col in df.columns], errors='ignore', inplace=True)
print(f"\nDropped original columns: {columns_to_drop}")


Dropped original columns: ['elements']


In [17]:
# Handle missing values for newly engineered features on df

newly_engineered_numeric_cols = [
    'ionic_radius_ion', 'atomic_weight_ion', 'charge_ion',
    'atomic_number_ion', 'electronegativity_ion',
    'e_fermi','nsites'
]

print("\nHandling missing values for newly engineered features and new MP features...")
for col in newly_engineered_numeric_cols:
    if col in df.columns:
        if df[col].isnull().any(): # Check if any NaN values are present in the column
            mean_val = df[col].mean()
            df[col] = df[col].fillna(mean_val)
            print(f"    Imputed NaN values in '{col}' with mean: {mean_val:.4f}")
        else:
            print(f"    '{col}' has no NaN values.")
    elif col not in df.columns: # For clarity if a listed column is entirely missing
        print(f"    Warning: Column '{col}' not found for NaN imputation.")


Handling missing values for newly engineered features and new MP features...
    'ionic_radius_ion' has no NaN values.
    'atomic_weight_ion' has no NaN values.
    'charge_ion' has no NaN values.
    'atomic_number_ion' has no NaN values.
    'electronegativity_ion' has no NaN values.
    'e_fermi' has no NaN values.
    'nsites' has no NaN values.


In [18]:
# Final check for NaNs

nan_counts_final = df.isnull().sum()
nan_counts_final = nan_counts_final[nan_counts_final > 0]

if not nan_counts_final.empty:
    print("\nWarning: NaN values still present in DataFrame after imputation (before final dropna):")
    print(nan_counts_final)
else:
    print("\nNo NaN values detected in the DataFrame after imputation (before final dropna).")



No NaN values detected in the DataFrame after imputation (before final dropna).


In [19]:
# Ensure original boolean columns are also explicitly cast to int
for col_prefix in ['is_metal', 'theoretical', 'working_ion']: # Check for both original name and new prefix for robustness
    # Check if the original boolean column still exists and is of boolean type
    if col_prefix in df.columns and df[col_prefix].dtype == 'bool':
        df[col_prefix] = df[col_prefix].astype(int)
        print(f"    Converted original '{col_prefix}' to int.")
    # Check if one-hot encoded versions are present (e.g., 'is_metal_0', 'is_metal_1')
    elif any(f"{col_prefix}_" in c for c in df.columns):
        print(f"    '{col_prefix}' or its related columns were one-hot encoded; no direct boolean conversion needed for original.")

    'is_metal' or its related columns were one-hot encoded; no direct boolean conversion needed for original.
    'theoretical' or its related columns were one-hot encoded; no direct boolean conversion needed for original.
    'working_ion' or its related columns were one-hot encoded; no direct boolean conversion needed for original.


In [20]:
# Final dropna applied to ensure all *modeling-relevant* features and targets have data.

features_to_exclude_from_dropna_subset = [
    'material_id', 'formula_pretty',
    'battery_id', # If this column exists and should be excluded from dropna subset
    'framework_formula', # If this column exists and should be excluded from dropna subset
    'discharged_formula' # If this column exists and should be excluded from dropna subset
]

In [21]:
# The subset for dropna should be all columns that are NOT excluded identifiers
dropna_subset = [col for col in df.columns if col not in features_to_exclude_from_dropna_subset]

initial_rows_before_final_dropna = len(df)
df.dropna(subset=dropna_subset, inplace=True) # Apply to df
print(f"Final dropna applied to ensure all *modeling-relevant* features and targets have data. New row count: {len(df)} (Dropped {initial_rows_before_final_dropna - len(df)} rows).")


Final dropna applied to ensure all *modeling-relevant* features and targets have data. New row count: 3445 (Dropped 0 rows).


In [26]:
# Now df is your final, engineered, clean DataFrame. Create df_engineered as a copy if desired.
df_engineered = df.copy()

print(f"DataFrame after Feature Engineering has {len(df_engineered)} rows and {len(df_engineered.columns)} columns.")
print("New features created/transformed, and all final features/targets are non-null.")
print("\n--- First 5 rows of the DataFrame after Feature Engineering: ---")
print(df_engineered.head().to_markdown(index=False, numalign="left", stralign="left")) # Use df_engineered here

DataFrame after Feature Engineering has 3445 rows and 53 columns.
New features created/transformed, and all final features/targets are non-null.

--- First 5 rows of the DataFrame after Feature Engineering: ---
| material_id   | energy_above_hull   | density   | formula_pretty   | nelements   | volume   | band_gap   | total_magnetization   | battery_type   | average_voltage   | theoretical_capacity_grav   | theoretical_capacity_vol   | energy_grav   | energy_vol   | num_steps   | max_voltage_step   | stability_charge   | stability_discharge   | e_fermi   | nsites   | num_unique_elements   | avg_atomic_number   | min_atomic_number   | max_atomic_number   | avg_atomic_mass   | min_atomic_mass   | max_atomic_mass   | avg_electronegativity   | min_electronegativity   | max_electronegativity   | avg_covalent_radius   | min_covalent_radius   | max_covalent_radius   | std_electronegativity   | ionic_radius_ion   | atomic_weight_ion   | charge_ion   | atomic_number_ion   | electronegativity_io

The DataFrame now contains 53 columns, which are categorized into identifiers, target variables, and input features for the machine learning model.

- 3 Identifier/Descriptive Columns: These columns (material_id, formula_pretty, battery_type) provide unique identification and descriptive information but are not used as direct inputs for the model.

- 3 Target Variables (Y): These are the properties our model will predict: theoretical_capacity_grav, average_voltage, and energy_above_hull.

- 47 Input Features (X): These columns represent the comprehensive set of material properties and characteristics used to train the model, comprising:

  - 14 Original Raw Features (retained/new from MP): These are properties directly acquired from the Materials Project that were kept for modeling. This now includes: density, nelements, volume, band_gap, total_magnetization, theoretical_capacity_vol, energy_grav, energy_vol, num_steps, max_voltage_step, stability_charge, stability_discharge, and the newly added e_fermi and nsites.
  - 14 Engineered Host Material Features: Derived from the host material's elemental composition (e.g., num_unique_elements, avg_atomic_number, max_electronegativity, std_electronegativity, min_covalent_radius, and others).
  - 5 Engineered Intercalating Ion Features: Properties specific to the working ion (e.g., ionic_radius_ion, atomic_weight_ion, charge_ion, atomic_number_ion, electronegativity_ion).
  - 14 Engineered One-Hot Encoded Features: Binary indicators for various categorical properties. This now includes:
    - 3 for the working ion (working_ion_Li, working_ion_Na, working_ion_Mg).
    - 7 for crystal_system (e.g., crystal_system_Cubic, crystal_system_Hexagonal, etc.).
    - 2 for is_metal (is_metal_False, is_metal_True).
    - 2 for theoretical (theoretical_False, theoretical_True).

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3445 entries, 0 to 3708
Data columns (total 53 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   material_id                  3445 non-null   object 
 1   energy_above_hull            3445 non-null   float64
 2   density                      3445 non-null   float64
 3   formula_pretty               3445 non-null   object 
 4   nelements                    3445 non-null   float64
 5   volume                       3445 non-null   float64
 6   band_gap                     3445 non-null   float64
 7   total_magnetization          3445 non-null   float64
 8   battery_type                 3445 non-null   object 
 9   average_voltage              3445 non-null   float64
 10  theoretical_capacity_grav    3445 non-null   float64
 11  theoretical_capacity_vol     3445 non-null   float64
 12  energy_grav                  3445 non-null   float64
 13  energy_vol             

## Save engineered data

In [24]:
# Save the DataFrame with engineered features to CSV

output_features_filename = '/content/drive/MyDrive/Colab Notebooks/Masters Research Project/Data/battery_materials_engineered_features_v2.csv'

try:
    df_engineered.to_csv(output_features_filename, index=False)
    print(f"\nDataFrame with engineered features saved to: '{output_features_filename}'")
except Exception as e:
    print(f"\nError saving engineered features DataFrame: {e}")


DataFrame with engineered features saved to: '/content/drive/MyDrive/Colab Notebooks/Masters Research Project/Data/battery_materials_engineered_features_v2.csv'


# Appendix: Detailed Feature Engineering Steps

This notebook performed a detailed feature engineering process on the battery materials data to prepare it for machine learning modeling. The objective was to transform the raw data into a richer set of predictive features by extracting and creating more meaningful numerical representations of both the host material and the intercalating ion, and to properly handle newly integrated materials properties.

Here's a breakdown of the key actions performed:

1.  **Data Loading**:
    - Loaded the raw battery materials data from the file `/content/drive/MyDrive/Colab Notebooks/Masters Research Project/Data/battery_materials_processed_v2.csv`.
    - The initial dataset contained 3709 rows and 25 columns.

2.  **Filter by Working Ion**:
    - Filtered the dataset to focus on common battery chemistries by keeping only entries where the `working_ion` is 'Li', 'Na', or 'Mg'.
    - This step dropped 264 rows, resulting in a DataFrame with 3445 rows.

3.  **Load Element Properties**:
    - Loaded comprehensive elemental properties from the PubChem CSV file located at `/content/drive/MyDrive/Colab Notebooks/Masters Research Project/Data/PubChemElements_all.csv`.
    - This data was used to create a lookup dictionary (`element_properties`) containing properties like atomic number, atomic mass, electronegativity, and covalent radius for each element.
    - Properties for 118 elements were successfully loaded. Note: 'AtomicRadius' from the CSV was used as a proxy for 'covalent_radius'.

4.  **Derived Host Material Features (14 Features)**:
    - For each material, features describing the host material's elemental composition were generated. This involved:
        - Parsing the string representation of the `elements` column into Python lists.
        - Calculating aggregate properties for the elements in each list using the loaded `element_properties` dictionary.
    - The 14 features created are:
        - `num_unique_elements`: The count of distinct elements in the material's formula.
        - Average, minimum, and maximum values for `atomic_number`.
        - Average, minimum, and maximum values for `atomic_mass`.
        - Average, minimum, and maximum values for `electronegativity`.
        - Average, minimum, and maximum values for `covalent_radius` (using AtomicRadius as a proxy).
        - `std_electronegativity`: The standard deviation of electronegativity for the elements.
    - A temporary column `elements_parsed` was used during calculation and then dropped.

5.  **Intercalating Ion Features (5 Features)**:
    - Properties specific to the `working_ion` (Li, Na, or Mg) were added to the DataFrame.
    - These features include:
        - `ionic_radius_ion`
        - `atomic_weight_ion`
        - `charge_ion`
        - `atomic_number_ion`
        - `electronegativity_ion`
    - These values were sourced from predefined properties (`ion_properties`) and supplemented with data from the `element_properties` dictionary.

6.  **One-Hot Encoded Categorical Features (14 Features)**:
    - Categorical columns were transformed into a numerical format suitable for machine learning models using one-hot encoding.
    - The columns encoded were:
        - `working_ion`: Resulting in `working_ion_Li`, `working_ion_Mg`, `working_ion_Na` (3 features).
        - `crystal_system`: Resulting in binary columns for each unique crystal system (e.g., `crystal_system_Cubic`, `crystal_system_Hexagonal`, etc.) (7 features).
        - `is_metal`: Resulting in `is_metal_False`, `is_metal_True` (2 features).
        - `theoretical`: Resulting in `theoretical_False`, `theoretical_True` (2 features).
    - One-hot encoding added a total of 14 new binary features to the DataFrame.
    - The shape of the DataFrame after one-hot encoding was (3445, 54).

7.  **Drop Redundant Columns**:
    - The original columns that were transformed or replaced by the feature engineering process were dropped to avoid redundancy and multicollinearity.
    - The dropped columns included: `elements`, `working_ion`, `crystal_system`, `is_metal`, and `theoretical`.

8.  **Handle Missing Values**:
    - Although the data from the Materials Project API was largely complete, a check and imputation step was performed for any potential missing values in the newly engineered numeric features and the newly added raw MP features (`e_fermi`, `nsites`).
    - Missing values in these specific columns were imputed using the mean of the respective column.
    - A final check confirmed that no NaN values were detected in the DataFrame after this imputation step.

9.  **Final Data Integrity Check**:
    - A final `dropna` operation was applied to the DataFrame, specifically targeting all columns *except* for identifier/descriptive columns (`material_id`, `formula_pretty`, `battery_type`).
    - This step ensured that every remaining data point had complete, non-null values for all features and target variables relevant for modeling.
    - 0 rows were dropped in this final step, confirming the completeness of the modeling-relevant data.

10. **Save Engineered Data**:
    - The final DataFrame (`df_engineered`) containing all the engineered and cleaned features was saved to a new CSV file: `/content/drive/MyDrive/Colab Notebooks/Masters Research Project/Data/battery_materials_engineered_features_v2.csv`.

**Outcome:**

The final DataFrame (`df_engineered`) has 3445 rows and 53 columns. These columns are categorized as:

-   **3 Identifier/Descriptive Columns**: `material_id`, `formula_pretty`, `battery_type`. These are not used as model inputs.
-   **3 Target Variables (Y)**: `theoretical_capacity_grav`, `average_voltage`, `energy_above_hull`. These are the properties to be predicted.
-   **47 Input Features (X)**: These are the features used to train the model and include:
    -   **14 Original Raw Features (retained/new from MP)**: `density`, `nelements`, `volume`, `band_gap`, `total_magnetization`, `theoretical_capacity_vol`, `energy_grav`, `energy_vol`, `num_steps`, `max_voltage_step`, `stability_charge`, `stability_discharge`, `e_fermi`, `nsites`.
    -   **14 Engineered Host Material Features**: Derived from elemental properties (e.g., `num_unique_elements`, `avg_atomic_number`, `max_electronegativity`, `std_electronegativity`, `min_covalent_radius`, etc.).
    -   **5 Engineered Intercalating Ion Features**: Properties of the working ion (`ionic_radius_ion`, `atomic_weight_ion`, `charge_ion`, `atomic_number_ion`, `electronegativity_ion`).
    -   **14 Engineered One-Hot Encoded Features**: Binary indicators for categorical properties (`working_ion_Li`, `working_ion_Mg`, `working_ion_Na`, `crystal_system_Cubic`, `crystal_system_Hexagonal`, `crystal_system_Monoclinic`, `crystal_system_Orthorhombic`, `crystal_system_Tetragonal`, `crystal_system_Triclinic`, `crystal_system_Trigonal`, `is_metal_False`, `is_metal_True`, `theoretical_False`, `theoretical_True`).

This comprehensive set of features is now ready for use in building and training machine learning models to predict battery material properties.